### ISSUE TITLE
Consider switching to the Dataframe API rather than hard-tethering to pandas

### ISSUE THREAD

**Issue by adamsardar** [lvgig/tubular#185]
**What**

[dataframe-api-compat](https://data-apis.org/dataframe-api-compat/) is an implementation of the [Python dataframe API standard](https://data-apis.org/dataframe-api/draft/index.html#index--page-root). This would permit tubular functions to operate on all dataframe libraries that implement the standard - most importantly [polars](https://pola.rs/).

**Why**

Currently the implementation of tubular is hard-tethered to pandas, and numpy-backed pandas at that for some cases (see explicit use of `np.int8` in [some functions](https://github.com/lvgig/tubular/blob/7b21a491de736a3fd9ee8069140e2809c349c478/tubular/imputers.py#L471)). This prevents users from adopting other libraries for analysis that might be better suited to their needs. In particular, the current implementation of tubular is quite memory hungry (see lvgig/tubular#182) and Polars permits so-called 'lazy-evaluation' and out-of-core computation.

**How**

Dataframe specific sections of tubular would need to be rewritten using the dataframe-agnostic API in [dataframe-api-compat](https://data-apis.org/dataframe-api-compat/basics/dataframe/). This is currently hard-blocked by the tests that inspect the way that functions are called (see lvgig/tubular#155), but with robust tests on behaviour we could look to extend support to polars. 

This item could be broken down into stages - transformers could be adapted one-by-one.

_References:_

- [Using Polars in a Pandas world](https://pythonspeed.com/articles/polars-pandas-interopability/)
- [Dataframe API Compat package documentation](https://data-apis.org/dataframe-api-compat/basics/dataframe/)
- [scikit-learn has some support for polars](https://scikit-learn.org/dev/modules/generated/sklearn.base.TransformerMixin.html#sklearn.base.TransformerMixin.set_output) as part of transform outputs.
  - See this scikit-learn/scikit-learn#25896 for some discussion on how sklearn are thinking about the polars/pandas support options.
- Also  look at data-apis/dataframe-api#346 for some insightful comments on the complexities of cross-DF support. It should make us question whether we really want to undertake such support ourselves.

**Comment by davidhopkinson26** [comment]
A nice idea.  Some thoughts on the transition period if transformers adapted one by one:

Do we want to accept a period where e.g. polars compatibility is only true for a subset of transformers (thus a pipeline must be constructed only using these ones, with polars -> pandas conversion happening outside that pipeline)

or 

do we as a first step want to add a framework for switching dataframe type within transformers as required.  This can eventually be removed once migration is complete.  Reading around it seems that this is not necessarily that big a slowdown assuming you are using pyarrow datatypes: [Manual in-memory conversion with .to_pandas](https://pythonspeed.com/articles/polars-pandas-interopability/)

Obviously the above only really addresses polars rather than anything coming through the dataframe agnostic API but that seems like the most likely application of this enhancement so the one most in need of management during the transition.

Seems we will likely want to roll back recent changes that downcast types which were somewhat ironically made in an attempt to improve speed!

The blocker of #155 should hopefully be resolved by #151 and #150 so I propose prioritising closing those before widespread implementation of https://data-apis.org/dataframe-api-compat/ but we could begin exploring it now.

**Comment by davidhopkinson26** [comment]
We should also think through handling of transformers we cannot migrate to the api.  Would be relatively simple to keep a method in BaseTransformer to convert to pandas for a set of expected input dataframe types and use this where needed across the package, but I'm not sure how to make this generic to anything in the standard.  That said, a set of likely formats is probably sufficient for our purposes.

**Comment by adamsardar** [comment]
I've had a look through the codebase with an eye on assessing how complex the transition would be. I think that it'd be ok - the best approach would be to create a upgrade branch and work through transform files in quick succession as PRs into that upgrade branch.

I think that it would actually simplify a fair chunk of our code, whilst also remove some code that makes assumptions over input (e.g. the use of np.nan - who knows how that's handled when input is in arrow dtype).

Whilst I suspect polars would be a commonly used type, pyarrow is also very popular.

The only methods that I don't think would map easily would be the pandas method transforms - they feel kind of hard to apply generically. I might be wrong though.

Agree that we should prioritise the other two tickets before cracking on with this, but the potential for polars support is too good not to explore.

**Comment by adamsardar** [comment]
A complication here comes from transforms that are hard coded to pandas methods, such as the [string transformer class](https://github.com/lvgig/tubular/blob/ec43ed457126c976afa06d650d238f959d5ed61f/tubular/strings.py#L96).

Will need to think about how to maintain support for pandas methods but also permit data frame agnostic calls.

**Comment by adamsardar** [comment]
I've been doing some more reading and thinking. Here is a summary:

- Polars support is the direction of travel in the data science community, and if that is a controversial statement then it is certainly going to be a popular option. See the linked discussion on sklearn for details, but I can easily imagine it eclipsing pandas in all but legacy applications over the next 12 months for performance reasons alone (I also prefer the syntax, but that is subjective).
- Preliminary benchmarking suggests that polars has comparable performance to operations on dictionaries, meaning that we would no longer need to maintain an in-house dict-backed version of tubular AND could use the same library for training and scoring; an obvious source of opportunities for different behaviour and bugs.
- Many workflows that we must support still use pandas (and will continue to use pandas for some time), therefore any support of polars must not be at a loss of pandas.
- [dataframe-api-compat](https://data-apis.org/dataframe-api-compat/) has good coverage across different dataframe libraries and general methods (`assign`, `drop` etc.).
  - I feel like the intended use is a general library that wants to support a wide range of dataframes.
- [narwhals](https://github.com/MarcoGorelli/narwhals) is a more minimal compatibility layer and has less coverage over general methods, but is tailored towards a polars way of doing things by using [expressions](https://docs.pola.rs/user-guide/concepts/expressions/). For a 
  - I feel like the intended use is a polars-backed library that wants to also support pandas, which is  more like our direction of travel.
  - There is the disclaimer `Note: this is work-in-progress, and a bit of an experiment, don't take it too seriously`. I'm not too worried about this seeing as it's on PyPI and an open-source project.
 

I had a look through tubular  and I think that most of our transformers could be represented as narwhals expressions. Those that can't (log transforms, string transforms) could either be fed into the OSS project as PRs or we could implement them in tubular via generic dispatch (i.e. have a method 'series_log` that can dispatch to a pandas version or a polars version appropriately). `Base`, `Capping`, `MeanResponse` etc. are all adaptable - I'd start there.

Bringing this ticket to a close then, I am proposing that we:
- expedite #150 and #151 as Dave suggests above.
- collate a list of the single-row methods that we want to address in the first tranche of transformers that we adapt.
- Create a PoC using narwhals on a branch and demonstrate a significant improvement of polars-backed frames over pandas-backed.
  - For methods that we doubt can ever be converted to a general dataframe method (such as the pandas-method transformer), consider accepting a compatability-layer frame and then casting [to_pandas()](https://marcogorelli.github.io/narwhals/api-reference/dataframe/).
  -  `narwhals` even has [an example](https://marcogorelli.github.io/narwhals/basics/complete_example/) of how to write a transformer class (clearly this is an intended use case).

**Comment by MarcoGorelli** [comment]
Hey - sorry to chime in, just wanted to say, thanks for taking a look at Narwhals! 🙏  From your comments, it seems you understand the spirit of the project perfectly!

A few comments:
- I'm not sure whether `dataframe-api-compat` is dead or whether it just smells funny. Since I left the project nearly 2 months ago, there's been exactly 0 commits, and the meetings have been cancelled due to lack of agenda items
- Narwhals is able to execute the first 5 TPC-H queries for pandas+Polars+cuDF, check https://www.kaggle.com/code/marcogorelli/narwhals-tpc-h-results-s-2/notebook and the related input notebooks. So, although there's still API coverage to add, I think it might already contain enough to be useful?
- Regarding the disclaimer: it's working out more nicely than I thought, so I think I'll remove it soon

Feel free to book some time on https://calendly.com/marcogorelli if you'd like to discuss how it could benefit you

**Comment by adamsardar** [comment]
Hi Marco,

Thanks so much for giving a bit of a sanity check on the thought process. It does feel as though narwhals is for people who want their (polars) cake and eat it at the same time.

I wasn't aware of the TPC-H benchmarks, nor your notebook - several of the queries look akin to our internal use, so that's positive.

> it's working out more nicely than I thought

Pleased to hear it! For what it's worth, I think that it's shaping up really nicely too.

I'll definitely take you up on that offer of a call - after easter. I just want to pull together an idea of which transforms we're prioritising in the first tranche of methods that we adapt.

**Comment by adamsardar** [comment]
From exploring some of the methods that we'd like to prioritise enabling polars support (via narwhals):

```
tubular.imputers.ArbitaryImputer
tubular.imputers.MedianImputer
tubular.imputers.ModeImputer

tubular.capping.CappingTransformer

tubular.nominal.GroupRareLevelsTransformer

tubular.mapping.MappingTransformer
```

(alongside their Base superclasses)

I'd probably start with `tubular.comparison.EqualityChecker` and `tubular.imputers.MeanImputer` (because they will likely be easiest) and then progress into the list above.

This ticket is getting a bit long in the tooth and I think that the relevant decision has been reached. I'll create a milestone for the above methods and ticket up the transformer adaptations so that we can share the load. Then I'll close this ticket with a link.

**Comment by adamsardar** [comment]
Closing this ticket in favour of #238.
